File: Grind 75 in Java/0155-min-stack.java
class MinStack {
    LinkedList<Pair<Integer, Integer>> stack = new LinkedList<>();
    
    public void push(int val) {
        int min = val;
        if (!stack.isEmpty())
            min = Math.min(min, getMin());
        stack.addLast(new Pair(val, min));
    }
    
    public void pop() {
        stack.removeLast();
    }
    
    public int top() {
        return stack.peekLast().getKey();
    }
    
    public int getMin() {
        return stack.peekLast().getValue();
    }
}

File: leetcode in kotlin/0155-min-stack.kt
class MinStack() {
    val stack = ArrayDeque<Pair<Int,Int>>() // value, minSoFar

    fun push(`val`: Int) {
        if(stack.isEmpty()) 
            stack.push(`val` to `val`)
        else
            stack.push(`val` to minOf(`val`, stack.peek().second))
    }

    fun pop() {
        stack.pop()
    }

    fun top(): Int {
        return stack.peek().first
    }

    fun getMin(): Int {
        return stack.peek().second
    }

}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = MinStack()
 * obj.push(`val`)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
